### DISCUSSION TITLE
The Future of the LRGS service

### DISCUSSION THREAD

**Discussion by MikeNeilson** [opendcs/opendcs#152] [Ideas]
@opendcs/opendcs-core-devs  @opendcs/opendcs-lrgs-devs 

After working with @cappelaere and digging through the LrgsMain, MsgArchive, and DDS source it's pretty clear that while it does currently work, there is a need for modernization.

# A brief description of the "LRGS" component.

This is specifically *what* the LRGS does, regardless of how. The what is also open for discussion.

- Aggregate Data from multiple sources, and source types, to a single archive of said raw data.
- Allow that archive to be streamed or queries to/from other LRGS systems.
- Current Input Sources are (i'm ignoring if defunct)
  - DRGS (dams-nt protocol)
  - HRIT/LRIT (dams-nt protocol)
  - NetworkDCPs (using DAMS-NT protocol)
  - NOAAport
  - Irridium (direct receive)
  - DomsetRecv
  - EDLInput
  - PolledNetworkDCP
  - ContinuousNetworkDCP
  - DDS (lrgs to lrgs streaming)
  - custom code (after some fixes)

DECODES, or specifically the routing specs or routing scheduler and various DECODES scripts work together to actually convert these to appropriate engineering units and send to the destination of choice, such as a database or files.

What I can current think of:

# General Modernization:

- Standardized Identity Management (LDAP, OpenID, OAuth). NOTE: self management of users still stays.
- SSL/TLS
- Initial configuration

# Servless (docker)

- Too many files vs a simple database connection, forces use of volumes in an odd way
- requires the thick client to do additional setup

While I think we could solve that within the current framework and DDS protocol, I'm strongly considering it might be worth just changing everything to be based on Jetty and creating a simple HTTP API and keeping everything in a database. SQLite would be used in cases where a large RDBMS is not desired or required.

Here's pro pitch:

- HTTP is a well standardized request/response protocol; you can get
- HTTP1.1 and HTTP2 have streaming protocol and event handlers
- Easier to setup strong authentication and security (SSL/TLS)
- The RtStat page is *already* an HTML page, maintaining that structure in JSP would be easier than raw java.

I don't generally go for "let's rewrite it", but after staring at the code all last weekend I think it would be both a similar amount of work to adjust the DDS protocol and LrgsMain to have external ID management and store the MsgArchive in a relational database. Not to mentioned possibly more disruptive.

# The actual proposal

Having dealt with data collection for many years, I very much like the idea of an extremely simple front end aggregator to do further work from. Anything more than store raw, gen basic stats, and move on is much more likely to fail or be complex. Thus I as a user of this tool do not desire anything beyond that for the LRGS service. However, one change I would agree with is being able to configure a destination for the raw data besides the archive itself.

So my thought is: Begin a new LrgsMain (Lrgs2 for the following text), in org/opendcs/lrgs. Use Jetty and Javalin (https://javalin.io/documentation) to start the user interaction API. Reuse, as practical, the current LrgsInput interfaces.

I'm open to tools other than javalin, but it's what I've used before and it has a good structure plus it can generate the API docs easily.

THe rough structure would likely be similar, except doubling down on what was clearly an intent and making all LrgsInputInterfaces and "MsgArchives" on equal footing.

The first milestone would be something that can be easily started with a few parameters (like dburl,username,password) and can be connected to and further setup.... I'm starting to ramble a little now, I'll likely come back and edit this later but I think this is getting most of my point across to get good feedback.

Bulk data transfer would be handled by a combination of HTTP/1.1 chunked encoding, HTTP/2 Data, websockets, and Server Sent Event as appropriate for a given user. But other protocols could be implemented, we want HTTP for simplicity in a lot of places but there can be a need for other things and we shouldn't forcibly limit ourselves.

**Comment by cappelaere** [comment]
It seems like you are talking about a client that could be used as an http API to the existing system.  This is something that I have also explored using nodejs and express.  On AWS, it would likely be implemented using the API Gateway and lambdas.  So many options there.  Are we sure that the core is solid enough?  Does it need refactoring?  I think we may all be in agreement.
Do we want to revisit the architecture using microservices?  How would this look like?  What features are missing or need to be refactored?  What would the user community like to see?

**Reply by MikeNeilson** [reply]
No, I'm talking about parallel replacement. The DDS protocol isn't difficult and I'm starting to better document it, but going forward I think HTTP does now provide support for everything we would need including sufficient performance. At least for the basic service itself. Support for older clients will likely be provided but the existing code just doesn't led itself easily to certain improvements, like moving the message archive to the SQL database or improved user management. That's all hardcoded in the existing protocol handlers.

As for microservices though, I don't think OpenDCS is going to go directly down that that route. But we should support the option for downstream users. Ignoring that there really isn't a good definition of microservices, OpenDCS is intended to be a set of tools that can be used in different ways. For example OLAV creates a industrial computer appliance that runs OpenDCS to do at Dam HRIT data collection and some initial processing for local operators. 

That said, given the way LRGS works it's pretty much always been a "microservice" and will likely stay that way.  I just try to avoid that name. If you are thinking of running it as a microservice that's a valid case and throwing out idea to help a certainly welcome. And you've already pointed a lot of that out; odd combination of setup xyz config file and then still need a thick client to fully setup the remaining options. Migrating the LRGS to use HTTP as it's communications mechanism can be done with minimal disruption to the end user's operations.

But my goal here is specifically: How can we simplify installation, deployment, and operation of the LRGS going forward and make it easier to use on cloud?

**Reply by cappelaere** [reply]
Several topics here.  Here are some comments:
1- Simplify installation. YES. We talked about izpack 5 to install without manual intervention
2- DDOS protocol update and cleanup.  It should not use SHA-1 anymore by default (unless specifically configured that way).  It should really interface with an identity component for authentication/authorization.  This component could have a default configuration (files) but also support some database or corporate single sign on as an option.
3- Client API. This is tricky to write a good one.  We are currently looking at different options.  HTTPS but also gRPC (which is 10x faster) and an event bus interface for cloud integration.  API itself might deserve to be designed with OpenAPI 3.x/GraphQL for synchronous operations and AsyncAPI for events
4- Cloud Implementation at the enterprise level requires more thought.  We need scalability, availability and high performance.  We need to be able to integrate with many of the best practice open source components that could give us critical features (Prometheus, OpenSearch, Grafana, TimeStreamDB...).  It will likely need to be deployed on clusters in multiple regions.  If this package gets redesigned, we may want to look at smaller components that are easy to customize/modify such as: data acquisition, data archiving, data distribution, logging, monitoring, computation, client... Very small services easy to deploy using docker/kubernetes.
5- Some organizations like us may be interested in data streaming and other web3.0 protocols to share data.

Before we embark on this adventure, it might behoove us to look at what Amazon does before starting a new project/product and how they focus on end users.  They call this: Working Backwards with creation of a Press release and FAQ.  Please Google it.  This is a tremendous opportunity to get end-users and upper management onboard and excited.  The purpose is to really generate new user stories and elicit requests for new capabilities that could be funded.
This does not have to be done immediately but we need a roadmap.  We need to use an agile development process that can reassure our respective organizations that we have a plan and are making progress.  How would you like to do this?
Looking forward for some feedback.
Cheers,
Pat.

**Reply by MikeNeilson** [reply]
1. I don't actually think that needs IZPack 5, just actually adding some values that now exist that haven't.
2. Totally agree, SHA1 shouldn't be used for anything security. I will update that as soon as I've been able to document the protocol appropriately and have written sufficient tests before making such a breaking change.
3. Not really, The focus of this thread is just the LRGS, and the needs are modest; by intention no processing is done on received messages, it's simply a "store-and-forward" system, for lack of a better term, the collects from multiple sources into a standard form. However, a strawman OpenAPI definition isn't a terrible idea.
4. NOTE: while the code is not broken up well, OpenDCS already is multiple modules that operate independently (other than some sharing a database). This thread is focused specifically on the LRGS component.  Version 8.0 is the start of breaking up the code base into more consumable modules.
5. I'm looking now at now HTTP/1.1 chunked data and WebSockets behave, those are well standarized protocols and will be prioritized for implementation. We're not oppose to new fangled things though; the plan is to be very modular so don't feel like that's shutting you down, just setting expectations for what I and any one who reports directly to me in some way may be working on.


The last paragraph:

@cappelaere  Metaphor time: The ship has literally left the port already, you are dinging catching up. (with good ideas, so keep them coming. Just to be clear there.)

Since the goal of the OpenDCS software is integrating with different system, we have enough to start embarking on a new version of the LRGS. It will be modular, the current system already had support for adding new Input interfaces, what it lacks is an easy to swap out storage backend and identity management (and modern monitoring). Theses are all fairly basic components including the monitor as we've already started adding JMX components to things.

The rest is beyond the scope of this discussion thread, please open a new one https://github.com/opendcs/opendcs/discussions/categories/development-practices if you would like to discuss such topics. That said, Amazon is never who I would look to as an example for critical infrastructure. Their risks are different, pretty much just money and are more than willing to risk loosing some to make some more. The OpenDCS software is used in water management operations in all of North and South America, and I suspect other continents as well; this can affect not just money, but life and property; life being the more important one there. We do things slowly and deliberately even if that means we're behind the times a little if things work and are stable. 

Our current role model is the Apache Software Foundation and projects like Tomcat. We already do work in an agile fashion, you are participating in that now. You've identified issues; hence posting this to the discussion forum. So far as I can tell most agree (some of this group tends to only chime in when they don't agree.) I care about this software but also cannot work on it full time nor do others involved.I do not believe the hype train suggested is necessary, nor beneficial. However; I am also not a dictator, open the discussion in the appropriate place above and consider taking a poll. I will concede to point if out voted by the other committers. You could also try pitching things to the general users mailing list https://www.freelists.org/list/opendcs.

That all said, *please* keep up the excitement, I'm just trying to direct it better in they way I think needs to be done to protect the project and our users.

**Reply by MikeNeilson** [reply]
if you do pitch that idea to the mailing list, please make your affiliations clear. 90% of that list will have no idea who you are or where your proposal is coming from, or why.

**Reply by cappelaere** [reply]
Regarding my affiliations:  I am current supporting the GOES-R port to the cloud (AWS) for NOAA/NESDIS.  We have received Authorization to Operate (ATO) for what we called the GeoCloud.  The current team is part of the NASA SES-III contract to support NOAA.  We have many GOES-R services to port to the Cloud as we are shutting down the processing on-premises. DCS and GRB are some of the services under consideration as we are also keeping an eye on GeoXO that will have much higher requirements.

**Comment by MikeNeilson** [comment]
Appropriate issues have been opened to track this effort. TLS has already been implemented and DDS-over-HTTP has been started.
